### ISSUE TITLE
Regression in 5.13: clang build fails with "undefined symbol: do_softirq"

### ISSUE THREAD

**Issue by v-fox** [ClangBuiltLinux/linux#1424]
5.12 was building fine with clang-12 but newly released 5.13 [fails with](https://build.opensuse.org/package/show/home:X0F:HSF:Kernel/kernel-source):
```
[ 6824s]   GEN     .version
[ 6824s]   CHK     include/generated/compile.h
[ 6824s]   GEN     .tmp_initcalls.lds
[ 6827s]   GEN     .tmp_symversions.lds
[ 6827s]   LTO     vmlinux.o
[ 7050s]   OBJTOOL vmlinux.o
[ 7057s]   MODPOST vmlinux.symvers
[ 7057s]   MODINFO modules.builtin.modinfo
[ 7057s]   GEN     modules.builtin
[ 7057s]   LD      .tmp_vmlinux.kallsyms1
[ 7058s] ld.lld: error: undefined symbol: do_softirq
[ 7058s] >>> referenced by fbdev.c
[ 7058s] >>>               vmlinux.o:(flush_smp_call_function_from_idle)
[ 7058s] >>> referenced by fbdev.c
[ 7058s] >>>               vmlinux.o:(netif_rx_ni)
```

Not making fbdev.c built-in only moves the same error to a different file.

**Comment by nathanchance** [comment]
Configuration and architecture?

**Comment by v-fox** [comment]
x86_64, config is in the above link as "[config.HSF](https://build.opensuse.org/source/home:X0F:HSF:Kernel/kernel-HSF/config.HSF?rev=844eb9f7dce61049bf03aadc27352c56)", here's copy: [config.txt](https://github.com/ClangBuiltLinux/linux/files/6736174/config.txt)

Note that CONFIG_PREEMPT_RT is force-enabled, that is usually not exposed in config menus. But I haven't seen it leading to build or boot errors yet:
```patch
diff --git a/arch/x86/Kconfig b/arch/x86/Kconfig
index 21f8511..b0b876d 100644
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -25,6 +25,7 @@ config X86_64
 	def_bool y
 	depends on 64BIT
 	# Options that are inherently 64-bit kernel only:
+	select ARCH_SUPPORTS_RT
 	select ARCH_HAS_GIGANTIC_PAGE
 	select ARCH_SUPPORTS_INT128 if CC_HAS_INT128
 	select ARCH_USE_CMPXCHG_LOCKREF
@@ -96,6 +97,7 @@ config X86
 	select ARCH_SUPPORTS_DEBUG_PAGEALLOC
 	select ARCH_SUPPORTS_NUMA_BALANCING	if X86_64
 	select ARCH_SUPPORTS_KMAP_LOCAL_FORCE_MAP	if NR_CPUS <= 4096
+	select ARCH_SUPPORTS_RT			if X86_64
 	select ARCH_USE_BUILTIN_BSWAP
 	select ARCH_USE_QUEUED_RWLOCKS
 	select ARCH_USE_QUEUED_SPINLOCKS
@@ -267,6 +269,9 @@ config STACKTRACE_SUPPORT
 config MMU
 	def_bool y
 
+config ARCH_SUPPORTS_RT
+	def_bool y if X86_64
+
 config ARCH_MMAP_RND_BITS_MIN
 	default 28 if 64BIT
 	default 8
```

**Comment by nickdesaulniers** [comment]
`do_softirq` is only defined in `kernel/softirq.c` when `CONFIG_PREEMPT_RT` is NOT defined.  Looks like those preprocessor guards were added in 8b1c04acad082dec76f3f8f7e1fa13493d6cbb79 which landed in 5.13-rc1.

**Comment by nickdesaulniers** [comment]
So does this work with GCC? I doubt this is a clang specific regression.

**Comment by v-fox** [comment]
> So does this work with GCC? I doubt this is a clang specific regression.

Haven't tried gcc yet as it takes 2 or more hours to check on OBS build-server and taking out my hacks, that force LLVM toolchain in build scripts, is cumbersome. Will try it out now.

**Comment by nathanchance** [comment]
Yeah this is not a clang specific regression so closing.

Apply the following diff to latest mainline:

```diff
diff --git a/arch/x86/Kconfig b/arch/x86/Kconfig
index 49ffb69e34dd..004b0bbe250e 100644
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -27,6 +27,7 @@ config X86_64
        # Options that are inherently 64-bit kernel only:
        select ARCH_HAS_GIGANTIC_PAGE
        select ARCH_SUPPORTS_INT128 if CC_HAS_INT128
+       select ARCH_SUPPORTS_RT
        select ARCH_USE_CMPXCHG_LOCKREF
        select HAVE_ARCH_SOFT_DIRTY
        select MODULES_USE_ELF_RELA
```

```
$ curl -LSso $(mkdir -p build/x86_64; echo build/x86_64)/.config https://github.com/ClangBuiltLinux/linux/files/6736174/config.txt

$ make -skj"$(nproc)" O=build/x86_64 olddefconfig vmlinux
ld: kernel image bigger than KERNEL_IMAGE_SIZE
ld: kernel/smp.o: in function `flush_smp_call_function_from_idle':
(.text+0xd24): undefined reference to `do_softirq'
ld: (.text+0xd29): undefined reference to `do_softirq'
ld: net/core/dev.o: in function `netif_rx_ni':
(.text+0x9f30): undefined reference to `do_softirq'
...
```

Feel free to report it upstream although the fact that no architecture selects `ARCH_SUPPORTS_RT` is probably telling that it is not ready for primetime.

**Comment by v-fox** [comment]
> Yeah this is not a clang specific regression so closing.

Sure, just finished a test build too.

That `kernel image bigger than KERNEL_IMAGE_SIZE` is even more annoying. Somehow I managed to get rid of it too but not sure how. Maybe CONFIG_CRASH_DUMP was at fault which I enabled by accident at some point. Or some locking related options, I think CONFIG_PROVE_LOCKING was broken in 5.12.

> Feel free to report it upstream although the fact that no architecture selects ARCH_SUPPORTS_RT is probably telling that it is not ready for primetime.

It's been on back-burner for years despite much more broken and less potentially useful features going in and out all the time. And since PREEMPT_RT isn't even advertised on x86_64, despite working for years even with 3rd-party modules like Nvidia GPU drivers, I doubt that upstream would be much interested. Thanks for the pointer, I would have never noticed that connection.
